### PR TITLE
[MM-19736] Inform clients what plugins getting show down when they get disabled

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
When plugins get disabled, we now send a websocket event to the clients informing them about the disabled plugins.

There are no unit tests for this code and it would be to much mana to add them in this ticket.

I've tested this PR locally and on the HA server. Still, please take a careful look as this is a quite important code pieces.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-19736

### CODE REVIEW

File: app/plugin.go [hunk 1]
```diff
@@ -111,8 +109,6 @@ func (a *App) SyncPluginsActiveState() {
 				}
 			}
 		}
-	} else { // If plugins are disabled, shutdown plugins.
-		pluginsEnvironment.Shutdown()
```
Comment (hanzei) [inline app/plugin.go]: I don't know why this was here in the first place. This should be done in `App.ShutDownPlugins()`

File: app/plugin.go [hunk 2]
```diff
@@ -551,9 +569,23 @@ func (a *App) notifyPluginEnabled(manifest *model.Manifest) error {
 	}
 
 	// Notify all cluster peer clients.
+	a.sendPluginEnableEvent(manifest)
+
+	return nil
+}
+
+func (a *App) sendPluginEnableEvent(m *model.Manifest) {
+	mlog.Debug("Sending plugin enable event", mlog.String("plugin id", m.Id))
```
Comment (hanzei) [inline app/plugin.go]: 1/5 that this helps with debugging.

File: app/plugin.go [hunk 3]
```diff
@@ -259,13 +255,35 @@ func (a *App) SyncPlugins() *model.AppError {
...
 		return
 	}
 
 	mlog.Info("Shutting down plugins")
 
+	availablePlugins, err := pluginsEnvironment.Available()
+	if err != nil {
+		mlog.Error("Failed to get available plugins while trying to shut them down. Clients will not get notified about shut down plugins.", mlog.Err(err))
+	}
+	for _, plugin := range availablePlugins {
```
Comment (crspeller) [inline app/plugin.go]: But doesn't this happen when any server in HA shuts down? You shut down one server and all your client plugins are disabled?

File: app/plugin.go [hunk 4]
```diff
@@ -259,13 +255,35 @@ func (a *App) SyncPlugins() *model.AppError {
...
+		pluginId := plugin.Manifest.Id
+		pluginEnabled := false
+		if state, ok := a.Config().PluginSettings.PluginStates[pluginId]; ok {
+			pluginEnabled = state.Enable
+		}
+
+		// Notifiy clients that the client has been shut down
+		if pluginEnabled && plugin.Manifest.HasClient() {
+			a.sendPluginDisableEvent(plugin.Manifest)
```
Comment (ali-farooq0) [inline app/plugin.go]: hmm does this work? I think `a.Srv.Hubs` are already stopped by this point as looking through the shutdown flow in this [function](https://github.com/mattermost/mattermost-server/blob/master/app/server_app_adapters.go#L157)

We'll have to move `s.FakeApp().ShutDownPlugins()` to the top for this to work, no?

**Comment by hanzei** [comment]
I moved the client shutdown code into the config handler. Please take a second look.

File: app/plugin.go (hunk 3 above)
Comment (hanzei) [reply app/plugin.go]: Good catch! Fixed.

Tested this case with the new code and can confirm that this doesn't happen now.

Comment (hanzei) [reply app/plugin.go]: Good catch! Fixed.

Tested this case with the new code and can confirm that this doesn't happen now.

**Comment by lieut-data** [comment]
@hanzei and I discussed this in more detail and identified a concern with my proposal. If it makes sense to do it for disabling plugins, it probably makes sense to do it for enabling plugins (which is also broken, and being addressed by a separate PR). But doing the latter would require all clients to turn around and make an API request to the server to fetch the activated plugin manifests -- potentially failing due to the coordinated slamming of a single endpoint.

At this point, I'm inclined to continue with @hanzei's current approach and rely on the existing websocket model to do both, even if there are some edge cases around HA and unusual plugin configurations.

File: app/plugin.go [hunk 5]
```diff
@@ -281,6 +282,36 @@ func (a *App) ShutDownPlugins() {
...
+	a.Srv.PluginsLock.RUnlock()
+
+	if pluginsEnvironment == nil {
+		mlog.Debug("Plugins are already shutdown while trying to inform clients about this")
+		return
+	}
+
+	availablePlugins, err := pluginsEnvironment.Available()
+	if err != nil {
+		mlog.Error("Failed to get available plugins while trying to shut them down. Clients will not get notified about shut down plugins.", mlog.Err(err))
```
Comment (lieut-data) [inline app/plugin.go]: Although it makes no difference for correctness, should we return early here for clarity?

Comment (hanzei) [reply app/plugin.go]: Sure :+1:

**Comment by hanzei** [comment]
If this approach has been approved in general, I will add a test for this.

**Comment by hanzei** [comment]
I'm going to close this PR. My code code changes introduced a regression where enabling the setting not longer also informs clients about the enabled plugins. I spend some mana trying to fix this, but didn't caught the root cause. Hence, I'm not confident in my changes and will defer this for someone else to work on.

**Pull request closed without merging**